backup: support SHOW BACKUPS with backup ids

kev-cao opened the pull request:
This commit teaches SHOW BACKUPS to display the new UX with backup IDs, along with BEFORE/AFTER filtering.

Epic: [CRDB-57536](https://cockroachlabs.atlassian.net/browse/CRDB-57536)

Resolves: https://github.com/cockroachdb/cockroach/issues/159647

Release note (sql change): Users can now set the `use_backups_with_ids` session setting to enable a new `SHOW BACKUPS IN` experience. When enabled, `SHOW BACKUPS IN <collection>` displays all backups in the collection. Results can be filtered by backup end time using `OLDER THAN <timestamp>` or `NEWER THAN <timestamp>` clauses.

Example usage:

    SET use_backups_with_ids = true;
    SHOW BACKUPS IN '<collection>' OLDER THAN '-1w' NEWER THAN '-2w';

kev-cao committed:
backup: support SHOW BACKUPS with backup ids

This commit teaches SHOW BACKUPS to display the new UX with backup IDs,
along with BEFORE/AFTER filtering.

Epic: CRDB-57536

Resolves: #159647

Release note (sql change): Users can now set the `use_backups_with_ids`
session setting to enable a new `SHOW BACKUPS IN` experience. When
enabled, `SHOW BACKUPS IN <collection>` displays all backups in the
collection. Results can be filtered by backup end time using `OLDER THAN
<timestamp>` or `NEWER THAN <timestamp>` clauses.

Example usage:

    SET use_backups_with_ids = true;
    SHOW BACKUPS IN '<collection>' OLDER THAN '2026-01-09 12:13:14' NEWER THAN '2026-01-04 15:16:17';

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/160137)
<!-- Reviewable:end -->

msbutler commented on pkg/sql/sessiondatapb/session_data.proto:
```diff
@@ -145,6 +145,11 @@ message SessionData {
   google.protobuf.Duration deadlock_timeout = 33 [(gogoproto.nullable) = false,
     (gogoproto.stdduration) = true];
+
+  // UseBackupsWithIDs, when true, enables the use of backup IDs in SHOW BACKUP
```
i think the commit could use a release note which explains how to use this new UX via the session var

msbutler commented on pkg/backup/show.go:
```diff
@@ -1353,6 +1354,70 @@ var showBackupsInCollectionHeader = colinfo.ResultColumns{
+	} else {
+		p.BufferClientNotice(
+			ctx,
+			pgnotice.Newf("no specific time range provided, defaulting to 24 hours"),
```
oh, i was under the impression that if the user specifies just `BEFORE X` (i.e. no AFTER), we return _all_ backups taken before X; and if they just specify `AFTER Y`, we return all backups after Y.... no 24 hour time window.

msbutler reviewed (commented):
Nearly there!

msbutler commented on pkg/backup/show.go:
```diff
@@ -1353,6 +1353,70 @@ var showBackupsInCollectionHeader = colinfo.ResultColumns{
 	{Name: "path", Typ: types.String},
 }
 
+var showBackupsWithIDsHeader = colinfo.ResultColumns{
```
looks like this leaked into the first commit

msbutler commented on pkg/backup/show.go:
```diff
@@ -1370,17 +1427,70 @@ func showBackupsInCollectionPlanHook(
+					pgnotice.Newf(
+						"only showing first %d backups. To see more, please run `SHOW BACKUPS IN ... OLDER THAN '%s'`",
+						len(res),
+						res[len(res)-1].EndTime.AsOfSystemTime(),
```
they can also specify both  OLDER THAN and NEWER THAN, right?

msbutler reviewed (commented):
nearly there!

kev-cao committed:
backup: update SHOW BACKUP syntax to NEWER/OLDER than

This commit updates the `SHOW BACKUP` time filtering syntax from
`AFTER/BEFORE` to `NEWER/OLDER THAN`.

Epic: CRDB-57536

Informs: #159647

Release note: None

kev-cao commented on pkg/backup/show_test.go:
```diff
@@ -827,3 +831,134 @@ func TestShowBackupCheckFiles(t *testing.T) {
+		// after the first backup. If the backup times overlap a second boundary,
+		// then our filter partitions the backups. Over the course of multiple
+		// tests, this gives us good coverage of filter points.
+		filterTime := int64(time.Duration(times[0]).Truncate(time.Second) + time.Second)
```
Oh yeah the comment is a bit misleading — I meant the second we choose is arbitrary, but since there's only really one or two seconds between the backups, it just ended up being the first second.

kev-cao commented on pkg/backup/BUILD.bazel:
```diff
@@ -133,6 +133,7 @@ go_library(
         "//pkg/util",
         "//pkg/util/admission",
         "//pkg/util/admission/admissionpb",
+        "//pkg/util/besteffort",
```
Hmm, yeah that's fair. I'm sad about it, it's so human-readable to just say "give me backups from 1 week ago".

kev-cao commented on pkg/backup/show_test.go:
```diff
@@ -827,3 +831,134 @@ func TestShowBackupCheckFiles(t *testing.T) {
+			)
+		})
+
+		t.Run("AFTER", func(t *testing.T) {
```
Added the test — after some more testing, you can sort of get some coverage if you just run the test enough times. So while _most_ cases don't really give you interesting results, it still gives us coverage over time.

msbutler commented on pkg/backup/backupinfo/backup_index.go:
```diff
@@ -342,39 +361,50 @@ func listIndexesWithinRange(
+				pendingEmit = nextEntry
+				return nil
+			}
+			if !nextEntry.end.After(pendingEmit.end) {
```
nit: I find "not after" a bit harder to reason about. To aid the reader, the comment could instead be written in a positive way:

"the pending emit has an end time less than or equal to the new entry. flush pending emit."

msbutler reviewed (approved):
Nice job! i've come around to your approach. thanks for entertaining my push back.

kev-cao commented on pkg/backup/backupinfo/backup_index.go:
```diff
@@ -248,56 +251,59 @@ type RestorableBackup struct {
+				// in descending order with ties broken by ascending start time, keeping
+				// the last one ensures that we keep the non-compacted backup.
+				if last.end.Equal(index.end) && last.fullEnd.Equal(index.fullEnd) {
+					*last = index
```
Actually, thinking about it some more, there's already test cases that cover this in `ListRestorableBackups`. I'll just add the test-only assertion.
